### PR TITLE
bump jquery to 3.0

### LINKED ISSUE #1215: Upgade jQuery dependency to 3.x
GitHub has security vulnerability against our 2.x version in packages/server/package.json. We should just upgrade to 3.x

screen shot 2018-01-22 at 12 38 13 pm

screen shot 2018-01-22 at 12 38 39 pm

### PR DESCRIPTION
**Pull request by jennifer-shehane**
- close #1215

### CODE REVIEW

**Comment by jennifer-shehane** [comment]
Definitely going to need to reference the upgrade guide for this: https://jquery.com/upgrade-guide/3.0/

**Comment by jennifer-shehane** [comment]
**First issue:** jQuery `.selector` has been removed in 3.0, which is used extensively to identify a node's selector. 

Example where `.selector` is used: 
- https://github.com/cypress-io/cypress/blob/issue-1215-jquery-3/packages/driver/src/cy/chai.coffee

**via jQuery**
> The property `.selector` was never a reliable indicator of the selector that could be used to obtain the set of elements currently contained in the jQuery set where it was a property, since subsequent traversal methods may have changed the set. 

> Plugins that need to use a selector string within their plugin can require it as a parameter of the method. For example, a "foo" plugin could be written as `$.fn.foo = function( selector, options ) { /* plugin code goes here */ };`, and the person using the plugin would write `$( "div.bar" ).foo( "div.bar", {dog: "bark"} );` with the `"div.bar"` selector repeated as the first argument of `.foo()`.

**Comment by Saibamen** [comment]
Please update your branch to latest develop branch from upstream.

I've tested all examples with JQuery 3.0.0 and only one test didn't pass:
`assertions.spec.js`: matches unknown text between two elements: `Timed out retrying: Expected to find element: 'undefined', but never found it.`

**Comment by Saibamen** [comment]
`.selector`s in `cypress\packages\driver`:
- src\cy\assertions.coffee: line 363
- src\cy\chai.coffee: lines 181, 232 and 252 but I think this is not JQuery
- src\cy\commands\querying.coffee: line 62
- src\cy\commands\traversals.coffee: line 50
- test\cypress\integration\commands\traversals_spec.coffee: line 155 and 157
- test\cypress\integration\commands\assertions_spec.coffee: test "uses $el.selector in expectation"

**Comment by bahmutov** [comment]
some tests are failing with undefined
```
   1) s1a "before each" hook for "t2a":
-     CypressError: Timed out retrying: Expected to find element: '.does-not-exist', but never found it.
+     CypressError: Timed out retrying: Expected to find element: 'undefined', but never found it.
```

**Comment by bahmutov** [comment]
You can recreate failing tests locally using for example
```
~/git/cypress/packages/driver on issue-1215-jquery-3
$ node ../../scripts/cypress run --project ./test \
	--spec test/cypress/integration/commands/querying_spec.coffee
```

**Comment by bahmutov** [comment]
with jQuery 2 if an element is not found we have selector property which we use to form error message

jquery2-not-found-selector


with jQuery 3 change we do not have this property (and we don't have context property)

jquery3-selector-not-found

**Comment by bahmutov** [comment]
Next missing test: `cy.wrap([1, 2]).first()` fails while creating a snapshot, because it tries to set attribute `data-cypress-el` on jQuery element `1`

```
TypeError: Cannot create property 'data-cypress-el' on number '1'
    at Function.prop (jquery.js:7472)
    at attr (jquery.js:7332)
    at access (jquery.js:3908)
    at jQuery.fn.init.attr (jquery.js:7310)
    at Object.createSnapshot (snapshots.coffee:120)
```

**Comment by bahmutov** [comment]
Snapshots CSS test failing because `$.load` was removed https://jquery.com/upgrade-guide/3.0/#breaking-change-load-unload-and-error-removed

```
1) driver/src/cy/snapshots_css .getStyleIds "before each" hook for "returns IDs for cached CSS contents":
     TypeError: url.indexOf is not a function

Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'driver/src/cy/snapshots_css'
```

Changing to the recommended solution `on('load', fn)`

**Comment by bahmutov** [comment]
Since snapshot test remains - the screenshot height is different from the snapshot value. See https://circleci.com/gh/cypress-io/cypress/150034

File: packages/runner/src/app/app.jsx [hunk 1]
```diff
@@ -152,11 +152,8 @@ class App extends Component {
       containerNode.className += ' screenshotting'
 
       if (!config.scale) {
-        const $window = $(window)
-        const $iframesSizeNode = $(iframesSizeNode)
-
-        iframesSizeNode.style.width = `${Math.min($window.width(), $iframesSizeNode.width())}px`
-        iframesSizeNode.style.height = `${Math.min($window.height(), $iframesSizeNode.height())}px`
+        iframesSizeNode.style.width = `${Math.min(window.innerWidth, iframesSizeNode.offsetWidth)}px`
```
Comment (chrisbreiding) [inline packages/runner/src/app/app.jsx]: jQuery 3 changes how it does width and height calculations (it now uses `getBoundingClientRect`), which takes into account the scaling. We want the absolute value without scaling, so we're better off directly using the DOM properties.

**Comment by jennifer-shehane** [comment]
@chrisbreiding I'm in favor of bumping it back down to 3.0.0 and pushing that out.

**Comment by chrisbreiding** [comment]
Was able to get 3.1.1 passing.

**Comment by jennifer-shehane** [comment]
Good team effort. Almost 2 years coming. 

@chrisbreiding I guess open a new PR for jQuery latest. Merging this one in.

**Comment by wKovacs64** [comment]
Yes! 🙌 Is a release imminent?

**Pull request merged**